collector/devstat_freebsd.go

```diff
@@ -0,0 +1,239 @@
+	uint64_t	write;
+	uint64_t	free;
+} Transfers;
+
+typedef struct {
+	double		other;
+	double		read;
+	double		write;
+	double		free;
+} Duration;
+
+typedef struct {
+	char		device[DEVSTAT_NAME_LEN];
+	int		unit;
+	Bytes		bytes;
+	Transfers	transfers;
+	Duration	duration;
+	long		busyTime;
+	uint64_t	blocks;
+} Stats;
+
+int _get_ndevs() {
+	struct statinfo current;
+	int num_devices;
+
+	current.dinfo = (struct devinfo *)calloc(1, sizeof(struct devinfo));
+	if (current.dinfo == NULL)
+		return -2;
+
+	devstat_checkversion(NULL);
+
+	if (devstat_getdevs(NULL, &current) == -1)
+		return -1;
+
+	return current.dinfo->numdevs;
+}
+
+Stats _get_stats(int i) {
+	struct statinfo current;
+	int num_devices;
+
+	current.dinfo = (struct devinfo *)calloc(1, sizeof(struct devinfo));
+	devstat_getdevs(NULL, &current);
+
+	num_devices = current.dinfo->numdevs;
+	Stats stats;
+	uint64_t bytes_read, bytes_write, bytes_free;
+	uint64_t transfers_other, transfers_read, transfers_write, transfers_free;
+	long double duration_other, duration_read, duration_write, duration_free;
+	long double busy_time;
+	uint64_t blocks;
+
+	strcpy(stats.device, current.dinfo->devices[i].device_name);
+	stats.unit = current.dinfo->devices[i].unit_number;
+	devstat_compute_statistics(&current.dinfo->devices[i],
+		NULL,
+		1.0,
+		DSM_TOTAL_BYTES_READ, &bytes_read,
+		DSM_TOTAL_BYTES_WRITE, &bytes_write,
+		DSM_TOTAL_BYTES_FREE, &bytes_free,
+		DSM_TOTAL_TRANSFERS_OTHER, &transfers_other,
+		DSM_TOTAL_TRANSFERS_READ, &transfers_read,
+		DSM_TOTAL_TRANSFERS_WRITE, &transfers_write,
+		DSM_TOTAL_TRANSFERS_FREE, &transfers_free,
+		DSM_TOTAL_DURATION_OTHER, &duration_other,
+		DSM_TOTAL_DURATION_READ, &duration_read,
+		DSM_TOTAL_DURATION_WRITE, &duration_write,
+		DSM_TOTAL_DURATION_FREE, &duration_free,
+		DSM_TOTAL_BUSY_TIME, &busy_time,
+		DSM_TOTAL_BLOCKS, &blocks,
+		DSM_NONE);
+
+	stats.bytes.read = bytes_read;
+	stats.bytes.write = bytes_write;
+	stats.bytes.free = bytes_free;
+	stats.transfers.other = transfers_other;
+	stats.transfers.read = transfers_read;
+	stats.transfers.write = transfers_write;
+	stats.transfers.free = transfers_free;
+	stats.duration.other = duration_other;
+	stats.duration.read = duration_read;
+	stats.duration.write = duration_write;
+	stats.duration.free = duration_free;
+	stats.busyTime = busy_time;
+	stats.blocks = blocks;
+
+	return stats;
+}
+*/
+import "C"
+
+const (
+	devstatSubsystem = "devstat"
+)
+
+type devstatCollector struct {
+	bytes     *prometheus.GaugeVec
+	transfers *prometheus.GaugeVec
+	duration  *prometheus.GaugeVec
+	busyTime  *prometheus.GaugeVec
+	blocks    *prometheus.GaugeVec
+}
+
+func init() {
+	Factories["devstat"] = NewDevstatCollector
+}
+
+// device stats.
+func NewDevstatCollector() (Collector, error) {
+	//var diskLabelNames = []string{"device"}
+	//var ioType = []string{"type"}
+
+	return &devstatCollector{
+		// Docs from https://www.kernel.org/doc/Documentation/iostats.txt
+		bytes: prometheus.NewGaugeVec(
```

Comment:
These values are calculated since boot time:
```
 The devstat_compute_statistics() function provides	complete statistics
     calculation.  There are four arguments for	which values must be supplied:
     current, previous,	etime, and the terminating argument for	the varargs
     list, DSM_NONE.  For most applications, the user will want	to supply
     valid devstat structures for both current and previous.  In some
     instances,	for instance when calculating statistics since system boot,
     the user may pass in a NULL pointer for the previous argument.  In	that
     case, devstat_compute_statistics()	will use the total stats in the
     current structure to calculate statistics over etime.
```
 And I think even "free" should be a counter:
```
     DSM_TOTAL_BYTES_READ

     DSM_TOTAL_BYTES_WRITE

     DSM_TOTAL_BYTES_FREE	     type: uint64_t *

				     The total number of bytes in transactions
				     of	the specified type between the acqui-
				     sition of previous	and current.

```

devstat(3): https://www.freebsd.org/cgi/man.cgi?query=devstat&sektion=3&apropos=0&manpath=FreeBSD+10.1-RELEASE